all.go

```diff
@@ -97,69 +101,112 @@ func runAll(
-	db, err := tsdb.Open(
-		storagePath,
-		logger,
-		prometheus.DefaultRegisterer,
-		&tsdb.Options{
-			RetentionDuration:      retention.Milliseconds(),
-			WALSegmentSize:         wal.DefaultSegmentSize,
-			MinBlockDuration:       tsdb.DefaultBlockDuration,
-			MaxBlockDuration:       retention.Milliseconds() / 10,
-			NoLockfile:             true,
-			AllowOverlappingBlocks: false,
-			WALCompression:         true,
-			StripeSize:             tsdb.DefaultStripeSize,
-		},
-	)
-	if err != nil {
-		return nil, err
-	}
+	lazyDBOpener := &lazyDB{}
+	scrapeManager := scrape.NewManager(log.With(logger, "component", "scrape-manager"), lazyDBOpener)
 
-	scrapeManager := scrape.NewManager(log.With(logger, "component", "scrape-manager"), db)
-
-	s, err := NewSampler(db, reloaders,
-		SamplerScraper(scrapeManager),
-		SamplerConfig(configFile),
-	)
-	if err != nil {
-		return nil, err
-	}
-	if err := s.Run(context.TODO(), g, reloadCh); err != nil {
-		return nil, err
-	}
-
-	w := NewWeb(mux, db, maxMergeBatchSize, queryTimeout,
+	w := NewWeb(mux, lazyDBOpener, maxMergeBatchSize, queryTimeout,
 		WebLogger(logger),
 		WebRegistry(reg),
 		WebReloaders(reloaders),
 		WebTargets(func(ctx context.Context) conprofapi.TargetRetriever {
 			return scrapeManager
 		}),
 	)
-	if err = w.Run(context.TODO(), reloadCh); err != nil {
+	if err := w.Run(context.TODO(), reloadCh); err != nil {
 		return nil, err
 	}
 
-	// run the grpc writable API
-	p, err = runStorage(
-		comp,
-		g,
-		p,
-		reg,
-		logger,
-		db,
-		srv.grpcBindAddr,
-		srv.grpcGracePeriod,
-		srv.grpcCert,
-		srv.grpcKey,
-		srv.grpcClientCA,
-	)
-	if err != nil {
-		return nil, err
-	}
+	go func() {
+		err := func() error {
+			db, err := tsdb.Open(
+				storagePath,
+				logger,
+				prometheus.DefaultRegisterer,
+				&tsdb.Options{
+					RetentionDuration:      retention.Milliseconds(),
+					WALSegmentSize:         wal.DefaultSegmentSize,
+					MinBlockDuration:       tsdb.DefaultBlockDuration,
+					MaxBlockDuration:       retention.Milliseconds() / 10,
+					NoLockfile:             true,
+					AllowOverlappingBlocks: false,
+					WALCompression:         true,
+					StripeSize:             tsdb.DefaultStripeSize,
+				},
+			)
+			if err != nil {
+
+				return err
+			}
+
+			lazyDBOpener.appender = db
+			lazyDBOpener.queryable = db
+
+			s, err := NewSampler(lazyDBOpener, reloaders,
+				SamplerScraper(scrapeManager),
+				SamplerConfig(configFile),
+			)
+			if err != nil {
+				return err
+			}
+			if err := s.Run(context.TODO(), g, reloadCh); err != nil {
+				return err
+			}
+
+			// run the grpc writable API
+			p, err = runStorage(
+				comp,
+				g,
+				p,
+				reg,
+				logger,
+				db,
+				srv.grpcBindAddr,
+				srv.grpcGracePeriod,
+				srv.grpcCert,
+				srv.grpcKey,
+				srv.grpcClientCA,
+			)
+			return err
+		}()
+		if err != nil {
+			level.Error(logger).Log("msg", "failed to start conprof components", "err", err)
+			os.Exit(2)
+		}
+	}()
 
 	p.Ready()
```

Comment:
We should only call ready when the DB is actually open.

Reply:
Updated! :+1: